### ISSUE TITLE
disengage.safety not long enough in case of not-responding minions

### ISSUE THREAD

**Issue by Martin-Weiss** [SUSE/DeepSea#962]
Just ran into this issue today:

admin-p:~ # salt-run disengage.safety; salt-run remove.osd 45
[WARNING ] Although 'dmidecode' was found in path, the current user cannot execute it. Grains output might not be accurate.
safety is now disabled for cluster ceph
[WARNING ] Although 'dmidecode' was found in path, the current user cannot execute it. Grains output might not be accurate.
[WARNING ] Although 'dmidecode' was found in path, the current user cannot execute it. Grains output might not be accurate.
False
[ERROR   ] ('Safety is not disengaged...refusing to remove OSD', ' run "salt-run disengage.safety" first THIS WILL CAUSE DATA LOSS.')
False

--> root cause seems to be that we have some minions registered to the salt-master on the admin that are not up and running and due to that do not respond..

admin-p:~ # salt '*' test-ping
admin-p.ses.intern.thomas-krenn.com:
    'test-ping' is not available.
osd04-p.ses.intern.thomas-krenn.com:
    'test-ping' is not available.
osd03-p.ses.intern.thomas-krenn.com:
    'test-ping' is not available.
osd01-p.ses.intern.thomas-krenn.com:
    'test-ping' is not available.
osd02-p.ses.intern.thomas-krenn.com:
    'test-ping' is not available.
sles12sp3-test3.ses.intern.thomas-krenn.com:
    Minion did not return. [No response]
sles12sp3-test1.ses.intern.thomas-krenn.com:
    Minion did not return. [No response]
sles12sp3-test4.ses.intern.thomas-krenn.com:
    Minion did not return. [No response]
sles12sp3-test2.ses.intern.thomas-krenn.com:
    Minion did not return. [No response]
ERROR: Minions returned with non-zero exit code

--> it seems that the "disengage.safety" and also the other jobs take too long in case one or the other minion is not reachable..

Is it possible to adjust the time the safety disengagement allows us to do "bad" things? 
Or could the processes somehow be configured to ignore the not available clients during these processes?

**Comment by jschmid1** [comment]
> admin-p:~ # salt '*' test-ping

it's `test.ping`

note the `.` instead of the `-`

> Is it possible to adjust the time the safety disengagement allows us to do "bad" things?

Currently it's hardcoded to 60 seconds, but can be changed if needed.

**Comment by Martin-Weiss** [comment]
Yeah - typo with the test-ping - but the problem stays the same ;-).

a) if salt minions are not reachable - deepSea commands are so slow... 
-> can this be solved somehow?

b) in case minions are so slow - could we add some parameter to the disengage.safety to allow a longer period "admin defined"?

**Comment by jschmid1** [comment]
> a) if salt minions are not reachable - deepSea commands are so slow...
> -> can this be solved somehow?

Right, that's because salt waits for a fixed amount of time for all it's minions to (re)connect. 

(with one minion down)
```
admin:~ # time salt-run validate.setup
[..]
real    2m47.822s
```

That exceeds the default duration of `disengage.safety` which suggests to increase the overall duration to about 5m~



> b) in case minions are so slow - could we add some parameter to the disengage.safety to allow a longer period "admin defined"?

I'll add that aswell.

**Comment by Martin-Weiss** [comment]
> 
> 
>     a) if salt minions are not reachable - deepSea commands are so slow...
>     -> can this be solved somehow?
> 
> Right, that's because salt waits for a fixed amount of time for all it's minions to (re)connect.
> 
> (with one minion down)
> 
> admin:~ # time salt-run validate.setup
> [..]
> real    2m47.822s
> 
> That exceeds the default duration of disengage.safety which suggests to increase the overall duration to about 5m~

And why does it matter if a minion is not reachable which is not even in the cluster?
(in this case it was a client minion)

**Comment by jschmid1** [comment]
> And why does it matter if a minion is not reachable which is not even in the cluster?
> (in this case it was a client minion)

because the default 'wait for minion to respond' time is exceeded by the disengage.safety duration now.
Salt doesn't care if your minion is part of the cluster or 'just' a client minion. It will just wait for it to come back before starting any operation.

**Comment by Martin-Weiss** [comment]
>     And why does it matter if a minion is not reachable which is not even in the cluster?
>     (in this case it was a client minion)
> 
> because the default 'wait for minion to respond' time is exceeded by the disengage.safety duration now.
> Salt doesn't care if your minion is part of the cluster or 'just' a client minion. It will just wait for it to come back before starting any operation.

Only if you address "all" minions and some are not reachable.
So is our targeting for disengage.safety or osd removal going to too many minions?

Addressing the relevant clients for an OSD removal:

```
admin-p:~ # time salt 'osd*' test.ping
osd02-p.ses.intern.thomas-krenn.com:
    True
osd03-p.ses.intern.thomas-krenn.com:
    True
osd01-p.ses.intern.thomas-krenn.com:
    True
osd04-p.ses.intern.thomas-krenn.com:
    True

real    0m0.536s
user    0m0.282s
sys     0m0.039s
```

Addressing all (including not relevant ones):
```
admin-p:~ # time salt '*' test.ping
admin-p.ses.intern.thomas-krenn.com:
    True
osd04-p.ses.intern.thomas-krenn.com:
    True
osd01-p.ses.intern.thomas-krenn.com:
    True
osd02-p.ses.intern.thomas-krenn.com:
    True
sles12sp3-test4.ses.intern.thomas-krenn.com:
    True
osd03-p.ses.intern.thomas-krenn.com:
    True
sles12sp3-test3.ses.intern.thomas-krenn.com:
    Minion did not return. [No response]
sles12sp3-test1.ses.intern.thomas-krenn.com:
    Minion did not return. [No response]
sles12sp3-test2.ses.intern.thomas-krenn.com:
    Minion did not return. [No response]

real    0m15.421s
user    0m0.493s
sys     0m0.034s

```

So it seems that the timeout problem only happens in case I address clients that are "out of scope" for my operation.

**Comment by jschmid1** [comment]
The difference between `salt-run` and a `salt` is that the former, indicating a _run_ner, runs on the _master_ and reaches out to minions from there on (or not, i.e. in the case of disengage.safety), whilst the latter, 'salt', queries the minions directly.

Salt-run somehow performs a check if all it's minions are up before invoking a runner. Haven't figured out if that can be disabled or at least configured to have a lower timeout.

So:

> Only if you address "all" minions and some are not reachable.
> So is our targeting for disengage.safety or osd removal going to too many minions?

No. That's not our fault.

> So it seems that the timeout problem only happens in case I address clients that are "out of scope" for my operation.

Right, again here the difference in invocation. salt vs salt-run

**Comment by Martin-Weiss** [comment]
> So:
> 
>     Only if you address "all" minions and some are not reachable.
>     So is our targeting for disengage.safety or osd removal going to too many minions?
> 
> No. That's not our fault.
> 
>     So it seems that the timeout problem only happens in case I address clients that are "out of scope" for my operation.
> 
> Right, again here the difference in invocation. salt vs salt-run

Hm - I can foresee that especially in bigger environments, bigger clusters and with many connected clients this will be a major problem and we really need to find a solution here. Any idea how we can solve that problem so that it does not matter if "out of scope" clients are causing long delays?

**Comment by swiftgist** [comment]
I think the resolution to this will be #1021.  If minions are not responding, the admin decides whether to fix those minions or remove them from the Salt cluster.  With the `salt-run manage.down removekeys=True`, that is a one liner.  When the nodes are working, then `salt-key -A` will add them back.

**Comment by Martin-Weiss** [comment]
> I think the resolution to this will be #1021. If minions are not responding, the admin decides whether to fix those minions or remove them from the Salt cluster. With the salt-run manage.down removekeys=True, that is a one liner. When the nodes are working, then salt-key -A will add them back.

With large environments and hundreds of minions registered to a given salt-master I do not think that removing and manually re-registering the minions is the best approach we can go.

An admin would have to keep track and have some "backlog" of the states executed while these minions were not reachable - and then take care of them once they are back manually and also find out if the not reachable ones are really dead or just the connection over the network lost for some time..

(keep in mind that there might be minions registered that are not deepsea related, too - thing of running deepsea on a SUSE manager salt-master etc..)
